src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
         if(!initialized)
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```

Comment:
Let me take a step back. It feels that the `ZooKeeperServer` object is really part of `ServerAuthenticationProvider` but not to the generic interface. Specifically, it feels like we need some interface that I'm going to call A, and both `AuthenticationProvider` and  `ServerAuthenticationProvider` specialize A.

The trouble of going down this path I'm suggesting is that we will need to make in the code to know whether we are dealing with one type or the other to determine which methods to call. For example, `matches` belongs in `AuthenticationProvider` while `matchesOp` belongs in `ServerAuthenticationProvider`.